contracts/utils/structs/MerkleTree.sol

```diff
@@ -0,0 +1,189 @@
+ * value, while keeping a history of merkle roots. This structure allows inserting commitment (or other entries) that
+ * are not stored, but can be proven to be part of the tree.
+ *
+ * The history of merkle roots allow inclusion proofs to remain valid even if leaves are inserted into the tree between
+ *  the moment the proof is generated and the moment it's verified.
+ *
+ * Each tree can be customized to use specific
+ * - depth
+ * - length of the root history
+ * - zero values (for "empty" leaves)
+ * - hash function
+ *
+ * WARNING:
+ *
+ * By design, the tree include zero leaves. Customizing the "zero value" might be necessary to ensure that empty leaves
+ * being provably part of the tree is not a security issue.
+ *
+ * _Available since v4.x._
+ */
+library MerkleTree {
+    /**
+     * @dev Maximum supported depth. Beyond that, some checks will fail to properly work.
+     * This should be enough for any realistic usecase.
+     */
+    uint256 private constant MAX_DEPTH = 255;
+
+    /**
+     * @dev Merkle tree cannot be initialized because requested depth is to large.
+     */
+    error MerkleTreeInvalidDepth(uint256 depth, uint256 maxDepth);
+
+    /**
+     * @dev The `sides` and `zero` arrays are set, at initialization, to have a length equal to the depth of the tree.
+     * No push/pop operations should be performed of these array, and their lengths should not be updated.
+     *
+     * The `roots` array stores the history of roots. Its length is set at initialization, and should not be updated.
+     *
+     * The hashing function used during initialization to compute the `zeros` values (value of a node at a given depth
+     * for which the subtree is full of zero leaves). This function is kept in the structure for handling insertions.
+     */
+    struct TreeWithHistory {
+        uint256 currentRootIndex;
+        uint256 nextLeafIndex;
+        bytes32[] sides;
+        bytes32[] zeros;
+        bytes32[] roots;
+        function(bytes32, bytes32) view returns (bytes32) fnHash;
+    }
+
+    /**
+     * @dev Initialize using the default hash
+     */
+    function initialize(TreeWithHistory storage self, uint256 depth, uint256 length, bytes32 zero) internal {
+        return initialize(self, depth, length, zero, MerkleProof.hashPair);
+    }
+
+    /**
+     * @dev Initialize a new complete MerkleTree defined by:
+     * - Depth `depth`
+     * - All leaves are initialize to `zero`
+     * - Hashing function for a pair of leaves is fnHash
+     * and keep a root history of length `length` when leaves are inserted.
+     */
+    function initialize(
+        TreeWithHistory storage self,
+        uint256 depth,
+        uint256 length,
+        bytes32 zero,
+        function(bytes32, bytes32) view returns (bytes32) fnHash
+    ) internal {
+        if (depth > MAX_DEPTH) {
+            revert MerkleTreeInvalidDepth(depth, MAX_DEPTH);
+        }
```

Comment:
Is it realistic to make the check? Not sure how costly will be to write to storage and perform a hash operation 255 times, but I wonder if it actually fits in a block.

Reply:
I think we need to put an arbitrary limit. I used 255, but really we would use anything "bug enough". 32 may be on the border. 64 would definitelly be enough for any real usecase.

Do you want to re-evaluate the limit ?

Reply:
No, I changed my main for two reasons:

1. 255 already allows 2**255 leaves, which is just before the `type(uint256).max`
2. 255 is `type(uint8).max`

I think we should treat `depth` the same way ERC20 treats `decimals` and restrict it to a `uint8`. I'll update